### PR TITLE
Token healing

### LINKED ISSUE #28346: Token healing (under 40 LOC)
### Feature request

Token healing rectifies the token boundary bias in greedy tokenization. It does this by trimming and regrowing the prompt to better align with the model's tokenizer, thus enhancing generation quality. The improvement is clearest with completion models.

Token boundary bias is a silent performance killer that doesn't seem very well known. It has clear impact on completion quality, though I'm not sure where it would fit as a transformers feature.

A more thorough explanation of the problem: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38).

### Motivation

Given a completion prompt with a partial url ending with `:`, the model might have seen the expected completion `://` as a _single_ token in training. However, the prompt's tail token `:` tells it that the next token is not `//`, and so it generates a wrong completion. Such errors compound in auto-regressive language models.

### Your contribution

My implementation (under 40 LOC): https://github.com/Ayenem/TokenHealer/tree/main

### PR DESCRIPTION
**Pull request by ahmed-moubtahij**
# What does this PR do?

Token healing rectifies the token boundary bias in greedy tokenization. It does this by trimming and regrowing the prompt to better align with the model's tokenizer, thus enhancing generation quality. The improvement is clearest with completion models.

Token boundary bias is a silent performance killer that doesn't seem very well known. It has clear impact on completion quality.

A more thorough explanation of the problem: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38).

### Motivation

Given a completion prompt with a partial url ending with `:`, the model might have seen the expected completion `://` as a _single_ token in training. However, the prompt's tail token `:` tells it that the next token is not `//`, and so it generates a wrong completion. Such errors compound in auto-regressive language models.

Fixes [#28346](https://github.com/huggingface/transformers/issues/28346)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

- @gante 
- @ArthurZucker

### CODE REVIEW

**Comment by Ayenem** [comment]
@ArthurZucker Follow-up on https://github.com/huggingface/transformers/pull/29081 :
* No more pygtrie dependency, I've added functionality to HF's Trie instead
* A -hopefully- prefix-robust space tokenization.

Tests failing on CI don't seem to be related to my code.

Please let me know if you see anything I can improve.

**Comment by Ayenem** [comment]
@LeonardoEmili Follow-up on https://github.com/huggingface/transformers/pull/29081

I have an example script on `examples/research_projects/token-healing/run_token_healing.py` let me know if that helps! Otherwise, please share which tokenizer you are using and which string the KeyError is raised on.

**Comment by Ayenem** [comment]
@gante @ArthurZucker CI is green :)

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -2884,6 +2885,42 @@ def test_return_unprocessed_logit_scores(self):
...
+class TokenHealingTestCase(unittest.TestCase):
+    @parameterized.expand(
+        [
+            (
+                "square_bracket",
+                'An example ["like this"] and another example [',
+                'An example ["like this"] and another example ["',
+            ),
+            ("url", 'The link is <a href="http:', 'The link is <a href="http://'),
+            ("aggressive_healing", 'The link is <a href="http', 'The link is <a href="http'),
```
Comment (gante) [inline tests/generation/test_utils.py:2899]: Why is this one called `agressive_healing`? The input and the expected result are the same, no?

**Review by gante — APPROVED** [review]
LGTM from the generate side 🙌 Thank you for iterating, and my apologies for the delayed review

nit: 
- two files are missing in the research project folder: the readme and the requirements file (which were present in the previous PR). Otherwise, our users won't really know what the project is about :)

File: tests/generation/test_utils.py (hunk 1 above)
Comment (ahmed-moubtahij) [reply tests/generation/test_utils.py:2875]: Thanks for getting back! I added back the two files and a comment explanation in the new push :)

**Comment by ArthurZucker** [comment]
Reviewing asap

File: src/transformers/tokenization_utils.py
Comment (ArthurZucker) [inline src/transformers/tokenization_utils.py:1]: Don't mind these changes in particular but we can have inheritance for the Trie to be a more specific trie without changing the more general Trie. 
TBH initializing on input tokens is nice

**Review by ArthurZucker — COMMENTED** [review]
Interesting work! 🔥

File: examples/research_projects/token-healing/run_token_healing.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+from transformers import AutoModelForCausalLM, AutoTokenizer, GenerationConfig
+
+
+def generate(inputs, model, tokenizer, token_healing):
+    input_ids = tokenizer(inputs, return_tensors="pt", padding=True).input_ids.cuda()
```
[the remark this answers is not in the document]
Comment (ahmed-moubtahij) [reply examples/research_projects/token-healing/run_token_healing.py:5]: You mean `input_ids = tokenizer(inputs, return_tensors="pt", padding=True).input_ids.to(torch_device)` ? I couldn't find where to import `torch_device` from

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1863,6 +1869,75 @@ def _has_unfinished_sequences(self, this_peer_finished: bool, synced_gpus: bool,
             return False
         return True
 
+    def heal_tokens(
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation/utils.py:1917]: @ArthurZucker custom pre-processing is in the plans for the `generate` refactor, and this should definitely be moved there (see step 2d in the [refactor plans](https://miro.com/app/board/uXjVNgMGfaQ=/), link only available to HF members for now). A separate file with extensions is a good idea!

Passing the tokenizer is an exception that we are accepting in other places -- e.g. see [this stopping criteria](https://github.com/huggingface/transformers/blob/e0c3cee17085914bbe505c159beeb8ae39bc37dd/src/transformers/generation/stopping_criteria.py#L139), which was added due to community request to stop generation in specific text strings.

In summary, token healing is "free performance" for generation and @ahmed-moubtahij did a lot of work with this PR. I would like to include it now and move it to a better location at refactor time, unless you heavily oppose 🤗

File: examples/research_projects/token-healing/run_token_healing.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply examples/research_projects/token-healing/run_token_healing.py:7]: `from transformers.testing_utils import torch_device`

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply examples/research_projects/token-healing/run_token_healing.py:7]: it's just to support any device. You can also use `device_map="auto"` to let it handle everything and no need to push the input ids to a device

**Comment by ArthurZucker** [comment]
feel free to ping me again for approval!

**Comment by ahmed-moubtahij** [comment]
> feel free to ping me again for approval!

@ArthurZucker It should be ok now :)

**Review by ArthurZucker — APPROVED** [review]
Thanks for getting this done! 
Left a few nits but good to go! 🔥

**Comment by ahmed-moubtahij** [comment]
@ArthurZucker Thank you for your review!
I could be wrong but the failing CI tests seem unrelated to my code. Is the convention to periodically fetch & rebase until it is fixed?

**Comment by ahmed-moubtahij** [comment]
CI passed :) @ArthurZucker @gante

**Comment by ArthurZucker** [comment]
Awesome! 🤗

**Comment by ArthurZucker** [comment]
I'll do a last pass now! 🤗

**Review by ArthurZucker — APPROVED** [review]

**Comment by ArthurZucker** [comment]
Thanks for you work @ahmed-moubtahij and congrats on the merge 🔥

**Comment by ahmed-moubtahij** [comment]
> Thanks for you work @ahmed-moubtahij and congrats on the merge 🔥

🥳🥳 Thank you and @gante for the assistance and follow-ups! I learned a lot :)
